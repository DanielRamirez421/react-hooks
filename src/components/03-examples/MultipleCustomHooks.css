body {
  padding: 70px;
}

.form-group, button {
  margin-top: 10px;
}

blockquote {
  text-align: right;
}