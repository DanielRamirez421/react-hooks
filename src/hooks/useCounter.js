import { useState } from "react";

export const useCounter = ( initialState = 10 ) => {
  
    const [state, setState] = useState(initialState);

    const increment = (factor) => setState( state + factor );
    const decrement = (factor) => setState( state > 0 ? state - factor : 0);
    const reset = () => setState( initialState );

    return {
        state,
        increment,
        decrement,
        reset
    }


    // Custom hook extract logical form a functional tool
};
