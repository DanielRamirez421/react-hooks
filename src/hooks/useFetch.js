import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  
  const isMounted = useRef(true); // Real life case to apps in production 
  const [state, setState] = useState({ data: null, loading: true, error: null });
    
  useEffect(() => {
    setState(s => ({ ...s, loading: true }));

    fetch( url )
      .then( resp => resp.json() )
      .then( data => isMounted && setState({ loading: false, error: null, data }));

    return () => isMounted.current = false;
  }, [url]);
  
  return state;
};
