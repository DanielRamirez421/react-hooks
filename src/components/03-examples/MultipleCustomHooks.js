import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from './../../hooks/useCounter';

import './MultipleCustomHooks.css'

export const MultipleCustomHooks = () => {

  const { state:counter, increment, decrement } = useCounter(1);
  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${ counter }`);
  const { author, quote } = !!data && data[0];

  return (
  <div>
    <h1>Breaking bad Quotes</h1>
    <hr />

    { loading ? 
        <div className='alert alert-info text-center'>Loading ...</div> :
        <blockquote className='text-right blockquote'>
          <p className='mb-3'>{ quote }</p>
          <footer className='blockquote-footer'>{ author }</footer>
        </blockquote>      
    }

    
    { counter !== 1 && <button 
                          className='btn btn-primary me-3' 
                          onClick={ () => decrement(1) }
                          >Previous Quote: {counter - 1 }
                        </button>}
    <button 
      className='btn btn-primary' 
      onClick={ () => increment(1) }
      >Next Quote: {counter + 1 }
    </button>

  </div>
  );
};
