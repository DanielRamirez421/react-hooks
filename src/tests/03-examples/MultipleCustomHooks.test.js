import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

jest.mock('../../hooks/useCounter');
jest.mock('../../hooks/useFetch');

describe('MultipleCustomHooks tests', () => {

  beforeEach(() => {
    useCounter.mockReturnValue({
      state: 10,
      counter: 10,
      increment: () => {}
    });
  });

  test('Should show component', () => {
    useFetch.mockReturnValue({ data: null, loading: true, error: null });
    const wrapper = shallow( <MultipleCustomHooks /> );
    expect(wrapper).toMatchSnapshot();
  });

  test('should show info in component', () => {
    const quote = 'HelloWorld';
    const author = 'Daniel';
    useFetch.mockReturnValue({ data: [{ author, quote }], loading: false, error: null });

    const wrapper = shallow( <MultipleCustomHooks /> );
    expect(wrapper.find('.alert-info').exists()).toBeFalsy();
    expect(wrapper.find('.mb-3').text().trim()).toBe(quote);
    expect(wrapper.find('.blockquote-footer').text().trim()).toBe(author);
  });


});
