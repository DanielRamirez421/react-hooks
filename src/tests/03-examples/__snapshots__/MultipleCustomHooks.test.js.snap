// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks tests Should show component 1`] = `
<div>
  <h1>
    Breaking bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-primary me-3"
    onClick={[Function]}
  >
    Previous Quote: 
    9
  </button>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote: 
    11
  </button>
</div>
`;
